lib/streamlit/components/v1/base_component_registry.py

```diff
@@ -0,0 +1,75 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+
+from streamlit import util
+from streamlit.components.v1.custom_component import CustomComponent
+
+
+class BaseComponentRegistry(ABC):
```

Comment:
We don't have strict rules here, but some of our "base" classes are inherited from `typing.Protocol` (they play nice with `@abstractmethod` decorator https://peps.python.org/pep-0544/#protocol-members ) and others from ABC.

Reply:
Ah great point, thanks. I have read about `Protocol` and as a consequence changed the `BaseComponentRegistry` to inherit from `Protocol`!